pkg/controller/activeprobe/activeprobe_controller.go

```diff
@@ -0,0 +1,292 @@
+		return
+	}
+
+	// TODO should support timeout and max retry
+	if err := r.syncActiveProbe(objKey); err == nil {
+		klog.Errorf("sync activeprobe  %v", objKey)
+		r.syncQueue.Forget(item)
+	} else {
+		klog.Errorf("Failed to sync activeprobe %v, error: %v", objKey, err)
+	}
+}
+
+func (r *Reconciler) syncActiveProbe(objKey k8stypes.NamespacedName) error {
+	var err error
+	ctx := context.Background()
+	ap := activeprobev1alph1.ActiveProbe{}
+	if err := r.Get(ctx, objKey, &ap); err != nil {
+		klog.Errorf("unable to fetch activeprobe %s: %s", objKey, err.Error())
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return client.IgnoreNotFound(err)
+	}
+
+	switch ap.Status.State {
+	case "":
+		err = r.runActiveProbe(&ap)
+	case activeprobev1alph1.ActiveProbeRunning:
+		err = r.checkActiveProbeStatus(&ap)
+	case activeprobev1alph1.ActiveProbeFailed:
+		r.deallocateTagForAP(&ap)
+	default:
+	}
+
+	return err
+}
+
+func (r *Reconciler) AddEndpointInfo(ap *activeprobev1alph1.ActiveProbe) error {
+	srcEpExternalIDValue := ap.Spec.Source.Endpoint
+	srcEndpointID := ctrltypes.ExternalID{
+		Name:  endpointExternalIDKey,
+		Value: srcEpExternalIDValue,
+	}
+	dstEpExternalIDValue := ap.Spec.Destination.Endpoint
+	dstEndpointID := ctrltypes.ExternalID{
+		Name:  endpointExternalIDKey,
+		Value: dstEpExternalIDValue,
+	}
+
+	var epList = securityv1alpha1.EndpointList{}
+	err := r.Client.List(context.Background(), &epList)
+	if err != nil {
+		klog.Errorf("list endpoint: %s", err)
+		return nil
+	}
+	for _, ep := range epList.Items {
+		if ep.Spec.Reference.ExternalIDName == srcEndpointID.Name && ep.Spec.Reference.ExternalIDValue == srcEndpointID.Value {
+			ap.Spec.Source.IP = ep.Status.IPs[0].String()
+			ap.Spec.Source.MAC = ep.Status.MacAddress
+			ap.Spec.Source.AgentName = ep.Status.Agents[0]
+			ap.Spec.Source.BridgeName = ep.Status.BridgeName
+			ap.Spec.Source.Ofport = ep.Status.Ofport
+		}
+		if ep.Spec.Reference.ExternalIDName == dstEndpointID.Name && ep.Spec.Reference.ExternalIDValue == dstEndpointID.Value {
+			ap.Spec.Destination.IP = ep.Status.IPs[0].String()
+			ap.Spec.Destination.MAC = ep.Status.MacAddress
+			ap.Spec.Destination.AgentName = ep.Status.Agents[0]
+			ap.Spec.Destination.BridgeName = ep.Status.BridgeName
+			ap.Spec.Destination.Ofport = ep.Status.Ofport
+		}
+	}
+	return nil
+}
+
+func (r *Reconciler) allocateTag(name string) (uint8, error) {
+	r.RunningActiveprobeMutex.Lock()
+	defer r.RunningActiveprobeMutex.Unlock()
+
+	for _, n := range r.RunningActiveprobe {
+		if n == name {
+			return 0, nil
+		}
+	}
+	for i := minTagNum; i <= maxTagNum; i += tagStep {
+		if _, ok := r.RunningActiveprobe[i]; !ok {
+			r.RunningActiveprobe[i] = name
+			return i, nil
+		}
+	}
+	return 0, fmt.Errorf("number of on-going ActiveProve operations already reached the upper limit: %d", maxTagNum)
+}
```

Comment:
RunningActiveprobe is stored in conntroller memory.
What will happen if controller HA to another one ?

Reply:
ok